scripts/data_overlap/ngram_hasher.py

```diff
@@ -0,0 +1,121 @@
+        self.window_start = 0
+        self.window_end = window_size
+        self.mod = mod
+        self.high_coefficient = pow(RABIN_KARP_PRIME, window_size - 1, mod)
+        self.current_hash = 0
+        self.int_sequence = int_sequence
+
+        for i in range(window_size):
+            self.current_hash = (self.current_hash * RABIN_KARP_PRIME + int_sequence[i]) % mod
+        if self.current_hash < 0:
+            self.current_hash += mod
+
+    def mult_mod(self, a: int, b: int, k: int) -> int:
+        return (a * b) % k
+
+    def update(self) -> None:
+        """
+        Update the rolling hash value when the window slides by one element.
+
+        - To efficiently update the hash:
+        1. Subtract the contribution of the element moving out of the window (start_piece).
+        2. Multiply the result by RABIN_KARP_PRIME.
+        3. Add the contribution of the new element entering the window.
+
+        Then update the window indices
+        """
+        start_piece = self.mult_mod(self.int_sequence[self.window_start], self.high_coefficient, self.mod)
+        self.current_hash = self.mult_mod(self.current_hash - start_piece, RABIN_KARP_PRIME, self.mod)
+        self.current_hash = (self.current_hash + self.int_sequence[self.window_end]) % self.mod
+        if self.current_hash < 0:
+            self.current_hash += self.mod
+        self.window_start += 1
+        self.window_end += 1
+
+
+def compute_hashes(int_sequence: List[int], window_size: int) -> List[Tuple[int, int, int]]:
+    """
+    Computes Rabin-Karp rolling hashes for a list of integers.
+
+    Args:
+        int_sequence (List[int]): List of integers.
+        window_size (int): Size of the rolling window.
+
+    Returns:
+        List[Tuple[int, int, int]]: List of tuples containing hash value, window start, and window end.
+    """
+    if len(int_sequence) < window_size:
+        return []
+    rk_hash = RabinKarpHash(int_sequence, window_size)
+    hashes = [(rk_hash.current_hash, rk_hash.window_start, rk_hash.window_end)]
+    for i in range(len(int_sequence) - window_size):
+        rk_hash.update()
+        hash_info = (rk_hash.current_hash, rk_hash.window_start, rk_hash.window_end)
+        hashes.append(hash_info)
+
+    return hashes
+
+
+# mod chosen as large prime of similar size to rabin karp, but distinct
+def hash_token(token: str, mod: int = 8554560727166512181):
+    """
+    Hashes a string token into an integer.
+
+    Args:
+        token (str): Input token.
+        mod (int): Modulus for hashing.
+
+    Returns:
+        int: Hash value.
+    """
+
+    # assume mod is constant for space/time efficiency
+    if token in hash_token_cache:
+        return hash_token_cache[token]
+
+    hash_value = 0
+    for ch in token:
+        hash_value = (hash_value * CHAR_PRIME + ord(ch)) % mod
+
+    hash_token_cache[token] = hash_value
+    return hash_value
+
+
+def get_ngram_hashes(tokens: List[str], n: int):
```

Comment:
To understand how much Rabin-Karp gives us over a baseline, suppose we replaced this function with `FarmHash64(" ".join(slice))` (from [this package](https://github.com/escherba/python-cityhash)) where `slice` is a sliding window of tokens:

- What is the time difference?
- What is the memory difference?

You could also use the standard library `hashlib` for this, but I would expect `hashlib` to be slower.